packages/react-router-dom/modules/Focus.js

```diff
@@ -0,0 +1,59 @@
+import React from "react";
+import { __RouterContext as RouterContext } from "react-router";
+import warning from "tiny-warning";
+
+class FocusWithLocation extends React.Component {
+  setRef = element => {
+    this.eleToFocus = element;
+  };
+
+  render() {
+    const { children } = this.props;
+    return children(this.setRef);
+  }
+
+  componentDidMount() {
+    this.focus();
+  }
+
+  componentDidUpdate(prevProps) {
+    // only re-focus when the location changes
+    if (this.props.location !== prevProps.location) {
+      this.focus();
+    }
+  }
+
+  focus() {
+    const { preserve, preventScroll = false } = this.props;
+    // https://developers.google.com/web/fundamentals/accessibility/focus/using-tabindex#managing_focus_at_the_page_level
+    if (this.eleToFocus != null) {
+      warning(
+        this.eleToFocus.hasAttribute("tabIndex") ||
+          this.eleToFocus.tabIndex !== -1,
+        'The ref must be assigned an element with the "tabIndex" attribute or be focusable by default in order to be focused. ' +
+          "Otherwise, the document's <body> will be focused instead."
+      );
+
+      if (preserve && this.eleToFocus.contains(document.activeElement)) {
+        return;
+      }
+
+      setTimeout(() => {
+        this.eleToFocus.focus({ preventScroll });
+      });
+    } else {
+      warning(
+        false,
+        "There is no element to focus. Did you forget to add the ref to an element?"
+      );
+    }
+  }
+}
+
+const Focus = props => (
+  <RouterContext.Consumer>
+    {context => <FocusWithLocation location={context.location} {...props} />}
+  </RouterContext.Consumer>
+);
+
```

Comment:
Should we add some `propTypes` to the `<Focus>` component here, like [we do elsewhere](https://github.com/ReactTraining/react-router/blob/5ccb241cd59eae7a15de31c7babff6619378be39/packages/react-router/modules/Redirect.js#L65-L71)?

I'm kinda on the fence about `propTypes` personally these days, since so much of the community is moving away from them and they just increase our bundle size. But I think at least for 4.x we should keep using them on public API.

Reply:
You can build your modules with a Babel plugin that strips prop types in production mode. That's what I do for RNWeb

Reply:
I meant to do this. This code was adapted from TypeScript, so I stripped out all of type references and then completely forgot to add prop types.

I would not miss prop types.

Reply:
Thanks for the tip, @necolas :) We used to use that plugin, but we use the `__DEV__` flag now instead, just so it's really clear when looking at the code which stuff is included in dev and which stuff isn't.

When I mentioned bundle size, what I meant was that I'm not sure if our `prop-types` import is able to be dropped in our production builds, because it isn't conditional. We aren't actually *using* prop types in our production code, so in theory it should be tree-shaked out of our production bundles, but we have a brand new bundling process in 4.4 (thx to @TrySound) and I still need to verify.